cmd/surf/commands/read.go

```diff
@@ -0,0 +1,130 @@
+package commands
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/CloudHub360/ch360.go/ch360"
+	"github.com/CloudHub360/ch360.go/config"
+	"github.com/CloudHub360/ch360.go/output/progress"
+	"github.com/CloudHub360/ch360.go/pool"
+	"io"
+	"os"
+)
+
+const ReadFilesCommand = "read"
+
+//go:generate mockery -name "FileReader|FilesProcessor"
+
+type FileReader interface {
+	Read(ctx context.Context, fileContent io.Reader, mode ch360.ReadMode) (io.ReadCloser, error)
+}
+
+type FilesProcessor interface {
+	RunWithGlob(ctx context.Context,
+		filesPattern string,
+		parallelism int,
+		processorFuncFactory ProcessorFuncFactory) error
+}
+
+type Read struct {
+	fileReader      FileReader
+	documentGetter  ch360.DocumentGetter
+	parallelWorkers int
+	filesProcessor  FilesProcessor
+	mode            ch360.ReadMode
+	filesPattern    string
+}
+
+func (cmd *Read) ProcessorFor(ctx context.Context, filename string) pool.ProcessorFunc {
+	return func() (interface{}, error) {
+		file, err := os.Open(filename)
+		if err != nil {
+			return nil, wrapErr(filename, err)
+		}
+
+		readCloser, err := cmd.fileReader.Read(ctx, file, cmd.mode)
+
+		return readCloser, wrapErr(filename, err)
+	}
+}
+
+func wrapErr(filename string, err error) error {
+	if err == nil {
+		return nil
+	}
+	return errors.New(fmt.Sprintf("Error reading file %s: %v", filename, err))
+}
+
+func (cmd *Read) Execute(ctx context.Context) error {
+
+	// Get the current number of documents, so we know how many slots are available
+	docs, err := cmd.documentGetter.GetAll(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Limit the number of workers to the number of available doc slots
+	parallelWorkers := min(cmd.parallelWorkers, ch360.TotalDocumentSlots-len(docs))
+
+	return cmd.filesProcessor.RunWithGlob(ctx,
+		cmd.filesPattern,
+		parallelWorkers,
+		cmd)
+}
+
+func NewReadFilesCommand(reader FileReader,
+	processor FilesProcessor,
+	mode ch360.ReadMode,
+	filesPattern string,
+	parallelWorkers int,
+	getter ch360.DocumentGetter) *Read {
+	return &Read{
+		filesProcessor:  processor,
+		fileReader:      reader,
+		documentGetter:  getter,
+		parallelWorkers: parallelWorkers,
+		filesPattern:    filesPattern,
+		mode:            mode,
+	}
+}
+
+func NewReadFilesCommandFromArgs(params *config.RunParams, client *ch360.ApiClient) (*Read, error) {
+
+	progressHandler, err := progress.NewProgressHandlerFor(params, os.Stderr)
+
+	if err != nil {
+		return nil, err
+	}
+
+	filesProcessor := &ParallelFilesProcessor{
+		ProgressHandler: progressHandler,
+	}
+
+	var readMode = ch360.ReadText
+	if params.ReadPDF {
+		readMode = ch360.ReadPDF
+	} else if params.ReadWvdoc {
+		readMode = ch360.ReadWvdoc
+	}
+
+	// ensure we're not printing binary data to the console
+	if !config.IsOutputRedirected() &&
+		readMode.IsBinary() &&
+		!params.IsOutputSpecified() {
+		return nil, errors.New("Not printing binary data to the console. Did you mean to use '-o'?")
```

Comment:
✋ I don't think this error is clear enough. It doesn't clearly explain what caused the error. I think it would be ok to ignore wvdocs for the purpose of this error (given I wasn't going to document them anyway). I suggest something like "You must use the --output-file (-o) parameter when the output format is pdf."